test/assertions_test.rb

```diff
@@ -316,40 +316,57 @@ def test_nested_assertions
     end
   end
 
-  def test_assertion_with_exceptions
+  def test_assertion_block_with_expected_exceptions
     assert_no_assertion_triggered do
-      @test_case.assert_raises(RuntimeError) do
-        @test_case.assert_statsd_increment('counter') do
-          StatsD.increment('counter')
-          raise "foo"
+      @test_case.assert_statsd_increment('expected_happened') do
+        @test_case.assert_raises(RuntimeError) do
```

Comment:
Since the exception is rescued before getting to the code under test, does this half of the test actually test a code path that isn't tested elsewhere?

Reply:
From a code coverage point of view, you could rewrite this as:
``` ruby
@test_case.assert_statsd_increment('expected_happened') do
  StatsD.increment('expected_happened')
end
```
For that pattern we already have tests for elsewhere, so we could :fire: this from a code coverage point of view.

However, I see this test more as documentation. The exception handling behavior of these assertion methods are not very intuitive, which explains why we keep tweaking these methods. So being a bit more explicit about how things will behave is a good thing.

Reply:
If it is documentation, then I don't think this is the most discoverable place for it.  We are currently missing documentation on `assert_statsd_increment`, so we could actually document it and show this patterns as an example in that documentation.  Alternatively, we could provide the example in the failure message for when an unexpected exception occurs inside an `assert_statsd_increment` block.

I think the exception handling behaviour in this case is intuitive.  Without knowing anything about this gem, a ruby developer could tell that

```ruby
        assert_raises(RuntimeError) do
          begin
            raise "expected"
          rescue
            StatsD.increment('expected_happened')
            raise
          end
        end
```

would be equivalent to `StatsD.increment('expected_happened')`, unless `StatsD.increment('expected_happened')` raises another exception.  That seems like a convoluted way of testing that though, so also doesn't serve well as an example of how to test something.

Reply:
I like the example, because the `assert_statsd_increment` + `assert_raises` combination is an extremely common case in real application code (count the number of exceptions that occur), and where we exposed quite a lot of bugs since we shipped #166. 

The [README](https://github.com/Shopify/statsd-instrument#minitest) documents `assert_statsd_instrument` to some degree. We should really add YARD documentation to the assertion class.